pkg/apis/wksprovider/controller/wksctl/machine_actuator.go

```diff
@@ -406,41 +423,144 @@ func (a *MachineActuator) update(ctx context.Context, cluster *clusterv1.Cluster
 		contextLog.Info("Machine and node have matching plans; nothing to do")
 		return nil
 	}
-	nodeIsMaster := isMaster(node)
-	if nodeIsMaster {
-		if err := a.prepareForMasterUpdate(machine, node); err != nil {
+	log.Infof("........................NEW UPDATE FOR: %s...........................", machine.Name)
+	isMaster := isMaster(node)
+	if isMaster {
+		if err := a.prepareForMasterUpdate(); err != nil {
 			return err
 		}
-	} else if isUpOrDowngrade(machine, node) {
-		if err = a.checkAnyMasterNotAtVersion(machineVersion(machine)); err != nil {
-			return gerrors.Wrap(err, "Master nodes must be upgraded before worker nodes")
-		}
 	}
-	if err = a.performActualUpdate(installer, machine, node, nodePlan, c); err != nil {
-		// If the update of a master failed, reset the master label
-		if nodeIsMaster {
-			if seterr := a.setNodeLabel(node, masterLabel, ""); seterr != nil {
-				return gerrors.Wrapf(err, "Could not reset master label after failed update: %v", seterr)
+	upOrDowngrade := isUpOrDowngrade(machine, node)
+	contextLog.Infof("Is master: %t, is up or downgrade: %t", isMaster, upOrDowngrade)
+	if upOrDowngrade {
+		err = checkForVersionJump(machine, node)
+		if err != nil {
+			return err
+		}
+		version := machineutil.GetKubernetesVersion(machine)
+		nodeStyleVersion := "v" + version
+		originalNeedsUpdate, err := a.checkIfOriginalMasterNotAtVersion(nodeStyleVersion)
+		if err != nil {
+			return err
+		}
+		contextLog.Infof("Original needs update: %t", originalNeedsUpdate)
+		masterNeedsUpdate, err := a.checkIfMasterNotAtVersion(nodeStyleVersion)
+		if err != nil {
+			return err
+		}
+		contextLog.Infof("Master needs update: %t", masterNeedsUpdate)
+		isOriginal, err := a.isOriginalMaster(node)
+		if err != nil {
+			return err
+		}
+		contextLog.Infof("Is original: %t", isOriginal)
+		if (!isOriginal && originalNeedsUpdate) || (!isMaster && masterNeedsUpdate) {
+			return errors.New("Master nodes must be upgraded before worker nodes")
+		}
+		isController, err := a.isControllerNode(node)
+		if err != nil {
+			return err
+		}
+		contextLog.Infof("Is controller: %t", isController)
+		if isMaster {
+			if isController {
+				// If there is no error, this will end the run of this reconciliation since the controller will be migrated
+				if err := drain.Drain(node, a.clientSet, drain.Params{
+					Force:               true,
+					DeleteLocalData:     true,
+					IgnoreAllDaemonSets: true,
+				}); err != nil {
+					return err
+				}
+			} else if isOriginal {
+				return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, originalMaster)
+			} else {
+				return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, secondaryMaster)
 			}
 		}
+		return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, worker)
```

Comment:
question: Will this return cause an issue if the drain logic is running? i.e. do we need a different return in the ```if isController {``` block so this return isn't executed?